microprofile/WFLY-10711_health_check_extension.adoc

```diff
@@ -0,0 +1,99 @@
+
+* A https://github.com/jmesnil/wildfly-microprofile-health[prototype of a WildFly extension] for MicroProfile Health was developed during the development of MicroProfile Health 1.0.
+
+== Requirements
+
+* Support Eclipse MicroProfile Health 1.0 for applications deployed in WildFly.
+* Provide a management runtime operation to check the health of the application server and its deployed applications. The result of the runtime operation
+  will be the DRM representation of the https://github.com/eclipse/microprofile-health/blob/master/spec/src/main/asciidoc/protocol-wireformat.adoc#appendix-b-json-payload-specification[MicroProfile Health JSON payload]
+* Provide an optional HTTP endpoint for the https://github.com/eclipse/microprofile-health/blob/master/spec/src/main/asciidoc/protocol-wireformat.adoc[health wireformat].
+  If the HTTP endpoint is enabled, it requires the presence of the WildFly https://github.com/wildfly/wildfly-capabilities/blob/master/org/wildfly/management/http-interface/capability.adoc[HTTP management interface capability]
+* The HTTP endpoint is global to the server: it must aggregate health check probes from *all deployments*.
+
+== Nice-to-Have Requirements
+
+* Support drilldown of the HTTP endpoint (e.g. to be able to query the health of individual deployment, e.g. `/health/deploymentA`)
+* Provide a WildFly management API so that other extensions can define health check probes that taken into account
+  by the HTTP endpoint (this is out of scope of this proposal but the design of the new extension must not preclude to
+  provide such an API at a later point).
+
+== Non-Requirements
+
+* MicroProfile Health 1.0 only provides a "liveness" probe (as defined by https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#define-a-liveness-command[Kubernetes]).
+__A Readiness probe is not covered by this proposal__.
+
+== Implementation Plan
+
+* Add dependencies to `io.smallrye:smallrye-health` and `org.eclipse.microprofile.health:microprofile-health-api:1.0` artifacts
+* Add a new `microprofile-health-smallrye` extension to provide Health support for deployments.
+* Provide a Undertow handler that exposes the optional health HTTP endpoint `/health` on WildFly `management-http` socket binding (port `9990` of the `management` network interface).
+
+=== Subsystem description
+
+The subsystem will be named `microprofile-health-smallrye`.
+It will define 1 runtime operation:
+
+* `check` whose result is the DRM representation of the Health Check JSON payload. If the global outcome is `UP`, the result's `outcome` will be `success` and the `result` will be the
+DMR representation of the HTTP JSON payload. Otherwise, the operation will fail.
+
+The subsystem will contains 2 attributes:
+* `http` - boolean to enable the HTTP endpoint on the HTTP management interface (false by default)
+* `http-endpoint` - String that represents the HTTP endpoint that exposes the health check (`/health` by default)
```

Comment:
I'm concerned about the complexity these combinations introduce. We need clients to be able to access this endpoint, and it needs to be easy to tell the client how to do so. Plus we should be cautious about assuming we will have control over the long term over this kind of client. Every config permutation we introduce has the potential to bite us.

Now a client needs to deal with these scenarios:

1) Health check subsystem isn't present in the system.
2) Health check subsystem is present in the system, but http=false so not exposed over http management.
3) Health check subsystem is present and exposed over http management, but they chose a different context.

We can write *our* client to work around the 2 and 3 cases by just using the DMR op, but then our client is bound to DMR (non-spec). But no general purpose client is likely to ever use the DMR op. Maybe this is ok, but it makes me uneasy versus just always exposing  this over http management on the 'health' context.

(Note that is http management is not enabled, other liveness and readiness checks that already exist will fail, as they rely on access to it. So we can assume http management exists. OTOH this means our overall liveness checks already are tied to DMR anyway, making my concern about using DMR for this even more theoretical.)

Anyone using the DMR op will need to distinguish the "health check subsystem isn't present in the system" case from other failures. I'll comment more about the DMR response separately.

The 'http-endpoint' attribute I don't see any need for right now, so long as we check with the JDG folks in case they are using that context for their uses of ExtensibleHttpManagement. We don't support use of that hook by anyone else, so we can take ownership of /health and remove any current reason to make it configurable when the management interface is used.

Is the 'http' attribute the way we would go if we ever wanted this exposed on 8080?

Reply:
`/health` is defined by MP spec, maybe we can relax configuration of the endpoint address.
Didn't check other app servers if they enable configuration of the endpoint address.

Reply:
> Is the 'http' attribute the way we would go if we ever wanted this exposed on 8080?

To have `/health` exposed both on 8080 and 9990 is not bad idea.
Users from other MP impls can be expecting to have `/health` on the same port as app server.

For example Thorntail exposes `http://localhost:8080/health`
https://github.com/thorntail/thorntail-examples/tree/master/jaxrs/health

Reply:
@bstansberry I see your points and we can make it simpler to configure.

* Let's mandate the presence of the http `management-interface` provided by [https://github.com/wildfly/wildfly-capabilities/blob/master/org/wildfly/management/http-interface/capability.adoc](org.wildfly.management.http-interface)  (so no more http enabled attributes or thing like that).
* Let's make the context always be `/health` (note that future version of the specification may add more context such as `/ready` but that's not decided at the moment). This context will be added to the http management interface by using the `org.wildfly.management.http.extensible` capability introduced by https://issues.jboss.org/browse/WFCORE-1742

I'm not sure if there is anything to do with regards to configuring http or https from the subsystem point of view.
It should only require the `org.wildfly.management.http-interface` capability. Whether the http management interface is exposed on HTTP or HTTPS is a concern of the [https://wildscribe.github.io/WildFly/13.0/core-service/management/management-interface/http-interface/index.html](core-service=management
management-interface=http-interface).

@bstansberry would that address your concern?

Reply:
@jmesnil Yes, this addresses my concern.  This basically leaves us with a DMR op, plus knowing that the /health context will be present on the http management interface if that interface exists. That meets the spirit of the spec (a /health context) and gives two solid options for using this in a liveness probe on our OpenShift image, while still leaving us with flexibility for the future if we get real end user requirements for other options (turn off http, expose it on 8080 etc.)

+1 to this work not concerning itself with http vs https on the management interface. All contexts on that interface are available over whatever protocol the overall interface config supports.